### ISSUE TITLE
[Multistore] - Product on category page display out of stock flag instead of availlable on order

### ISSUE THREAD

**Issue by Rizzen59** [PrestaShop/PrestaShop#27604]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

>Edited by Khouloud

Steps to reproduce the issue:
1. Enable Multistore & Create new Group shop with Share quantity = enabled
2. Create a shop assiciated to this Group shop
3. Create a product with quantity = 0 & allowed for order
4. Go to FO of the shop2 > Category list
5. See error > The product is displayed with out-of-stock flag
![image](https://user-images.githubusercontent.com/16067358/153606126-320f965a-e2e0-4302-b44a-9dd91f947656.png)

Hello, i'm back with my bug and i found the cause:

On catagory page i have product display as out of stock even if the product is availlable for order. But if I access to the product page, status is ok and product can be added on cart. More, when I add the product to cart, the status on category page change for the good status until the product are removed from the cart.

Not all product are affected. I searched in code and the problem come from the out_of_sotck parameter to define if you can order product out of stock. For some product the value of out_of_stock is empty on category page instead of 0,1 or 2. In this case it should be 1.

**Product.php line 5579:**

     ` $row['allow_oosp'] = Product::isAvailableWhenOutOfStock($row['out_of_stock']);`

Problem is only on category page.
Home page, product page, search page are ok.
To avoid loosing customer i disable message on category page

![123515636-7b7b8e00-d698-11eb-853f-76c3ce5b1a0c](https://user-images.githubusercontent.com/25029253/152675137-c44cd4ee-ffaf-4ccd-960f-0c635e5b574b.png)

![128632685-d1507f7f-9bf3-4c52-8459-34b62a81d6ac](https://user-images.githubusercontent.com/25029253/152675149-ebaed18e-23ac-48ef-9cbf-82baae25733f.png)

![128632731-46a0e0c9-5313-4dec-85c8-4040db14b05a](https://user-images.githubusercontent.com/25029253/152675154-59fd6ed0-5b2b-4ae7-8363-faadd7b23f43.png)

![128632843-5d73b17b-3530-48b5-939e-0583711376be](https://user-images.githubusercontent.com/25029253/152675158-4444a40f-40ce-4183-8e0b-edaba510dff0.png)

I try to come back on the classic theme to identify if the problem occurs too. At first glance, no message is displaying and it seems to be normal. But classic theme doesn't display out of stock or allows order by default. If stock equal 0, no message is displayed. I try with not allow and allow order:

**themes/classic/templates/catalog/_partials/product-flags.tpl**

```
{block name='product_flags'}
    <ul class="product-flags">
        {foreach from=$product.flags item=flag}
            <li class="product-flag {$flag.type}">{$flag.label}</li>
        {/foreach}
    </ul>
{/block}
```
But if i add the check on allow on order (product.allow_oosp) and display message, we can see that property is not initialize with correct value on catalog part and availability message is wrong :

**themes/classic/templates/catalog/_partials/product-flags.tpl**

```
{block name='product_flags'}
    <ul class="product-flags">
        {if $product.quantity > 0}
            {foreach from=$product.flags item=flag}
                <li class="product-flag {$flag.type}">{$flag.label}</li>
            {/foreach}
        {else}
            {if !$product.allow_oosp}
                <li class="product-flag out-of-stock">{$product.availability_message}</li>
            {/if}
        {/if}
    </ul>
{/block}
```

**Root cause :** 
After a long investigation, i found that : On class ProductAssemblerCore (ProductAssembler.php in classes foler), there is a SQL request that seach information for product. 
This request seach stock information by product shop and don't take account that stock is configurated to share stock 👍 

```
LEFT JOIN {$prefix}stock_available sa
			        ON sa.id_product = p.id_product
			        AND sa.id_shop = $idShop
```

![image](https://user-images.githubusercontent.com/25029253/152675471-d815e5f1-9383-4833-ae81-d25d5d5c89ac.png)

So on category page for shared stock management between  shop, information about stock are wrong

### Expected behavior

When a product is availlable for order even if its qty equal 0, display on demand on category page.

### Steps to reproduce

**You need to have multishop with shared product stock** 

1. Define product witch capability to be order even if stock are equal to 0
2. Define stock to 0
3. Show product on category page
4. Flag is out of stock instead of On-demand

### PrestaShop version(s) where the bug happened

1.7.8.3

### PHP version(s) where the bug happened

7.4

**Comment by hibatallahAouadni** [comment]
Hello @Rizzen59 

I didn't manage to reproduce the issue with PS1783, see the attached screen record below:

https://user-images.githubusercontent.com/23308427/152800584-5049e50b-9975-4acd-9ecd-627a8470c8ff.mp4

It seems this issue related to your overrides or some third-party modules.
Please check and feedback.

Thanks!

**Comment by Rizzen59** [comment]
It's seems you use a single shop not a multi shop with shared stock. 

![image](https://user-images.githubusercontent.com/25029253/152815847-43895024-e353-4d4c-923e-c3d44c5d6583.png)

On your video, I don't see the multishop header
![image](https://user-images.githubusercontent.com/25029253/152816141-e194518e-dc83-4524-913e-8b5cbc2efebb.png)

**Comment by Rizzen59** [comment]
The problem is in ProductAssemblerCore  that  is not an overided classe (ProductAssembler.php in classes foler)

There is a SQL request that seach information for product. This request seach stock information by product shop and don't take account that stock is configurated to share stock 👍

```
LEFT JOIN {$prefix}stock_available sa
			        ON sa.id_product = p.id_product
			        AND sa.id_shop = $idShop
```

In database multi shop with shared stock use the "id_shop_group" column, not the "id_shop" column to map product stock.
So that cannot work. :)

**Comment by Rizzen59** [comment]
Hello @hibatallahAouadni 

You can see the problem directly on the github code of prestashop  => ProductAssembler.php line 86 : 
https://github.com/PrestaShop/PrestaShop/blob/develop/classes/ProductAssembler.php

And the difference when sharing stock is used for exemple on ProductQueryBuilder.php line 184 : 
https://github.com/PrestaShop/PrestaShop/blob/116e6f596c79866aeb8bcc235a17020e2e5c2f21/src/Core/Grid/Query/ProductQueryBuilder.php

**Comment by khouloudbelguith** [comment]
Hi @Rizzen59,

Thanks for your feedback.
I manage to reproduce the issue with PS1783.
I attached a screen record

https://user-images.githubusercontent.com/16067358/153597750-2f9aaa0a-d54a-4ba1-b301-40a0f17f150c.mp4

The out of stock flag is a new behavior in PS1780
https://github.com/PrestaShop/PrestaShop/issues/21105

In the PS1778 => this flag is not displayed
so I can confirm that this behavior is indeed a bug.

Please be aware that due to the high number of reported issues, some bugs might take a very long time to be resolved. If this issue is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you. If you manage to get the problem fixed, then please contribute the fix back to the project as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thanks!

**Comment by Rizzen59** [comment]
Thanks for taking account this bug. 
If i have time, i will try to propose a solution. 

I have to find, how to transmit flag for enable shared stock and id_shop_group to ProductAssembler.php then update the sql about stock.

**Comment by MatShir** [comment]
@Rizzen59 you have done much of the investigation, you do not feel like contributing. We'll be happy to help you with the RP review. We are available on Slack OSS as well ;)

**Comment by Rizzen59** [comment]
@MatShir i'm looking how to contribute on prestashop and try to fix :)

**Comment by Rizzen59** [comment]
I'm working on fix,  i'll make the pull request this week

**Comment by Progi1984** [comment]
Fixed by @Rizzen59 in #27841 and tested by @sLorenzini
